### PR TITLE
AV-207705 Gateway API support for backendRef filters

### PR DESCRIPTION
**Pull request by pkoshtavmware**
AV-207705 Gateway API support for backendRef filters
It is done via datascript attached to the child vs via add/update/delete StringGroups

**TestingStatus:**
Unit Testcases added
EndToEnd testing for positive case done(CRUD)
Corner Case

**Testing pending:**
FT runs for sanity

### CODE REVIEW

**Comment by akshayhavile** [comment]
please correct the lint issue.

**Comment by pkoshtavmware** [comment]
> please correct the lint issue.

Done

File: internal/rest/dequeue_nodes.go [hunk 1]
```diff
@@ -108,6 +110,16 @@ func (rest *RestOperations) DequeueNodes(key string) {
 			rest.vrfCU(key, name, avimodel)
 			return
 		}
+		if strings.Contains(name, "StringGroup") {
```
Comment (akshayhavile) [inline internal/rest/dequeue_nodes.go:113]: `string group` and `datscript` functionality is related to EVH model only. So it should be handled in `RestOperationEVH`. We should have `name` enqueued as per that only. What is the reason of including this functionality here.

File: internal/cache/controller_obj_cache.go [hunk 2]
```diff
@@ -1765,6 +1771,92 @@ func (c *AviObjCache) PopulateL4PolicySetToCache(client *clients.AviClient, clou
...
 }
 
+func (c *AviObjCache) AviPopulateAllStringGroups(client *clients.AviClient, cloud string, StringGroupData *[]AviStringGroupCache, nextPage ...NextPage) (*[]AviStringGroupCache, int, error) {
+	var uri string
+	akoUser := lib.AKOUser
+
+	if len(nextPage) == 1 {
+		uri = nextPage[0].NextURI
+	} else {
+		uri = "/api/stringgroup/?" + "&include_name=true&created_by=" + akoUser
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply internal/cache/controller_obj_cache.go:1783]: Done

File: internal/cache/controller_obj_cache.go [hunk 3]
```diff
@@ -113,6 +115,10 @@ func (c *AviObjCache) AviRefreshObjectCache(client []*clients.AviClient, cloud s
 		defer wg.Done()
 		c.PopulateL4PolicySetToCache(client[6], cloud)
 	}()
+	go func() {
+		defer wg.Done()
+		c.PopulateStringGroupDataToCache(client[8], cloud)
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply internal/cache/controller_obj_cache.go]: Done

File: ako-gateway-api/nodes/avi_model_l7_translator.go [hunk 4]
```diff
@@ -371,3 +406,31 @@ func (o *AviObjectGraph) BuildHTTPPolicySetHTTPRequestRedirectRules(key string,
 		}
 	}
 }
+
+func (o *AviObjectGraph) ConstructBackendFilterDataScript(key string) *nodes.AviHTTPDataScriptNode {
+	datascripts := o.GetAviHTTPDSNode()
+	datascriptName := akogatewayapilib.Prefix + akogatewayapilib.BackendRefFilterDatascriptName
```
[the remark this answers is not in the document]
Comment (pkoshtavmware) [reply ako-gateway-api/nodes/avi_model_l7_translator.go]: Moved Naming function at lib folder with cluster name included and encoded as well.

File: internal/cache/controller_obj_cache.go [hunk 5]
```diff
@@ -113,6 +115,12 @@ func (c *AviObjCache) AviRefreshObjectCache(client []*clients.AviClient, cloud s
 		defer wg.Done()
 		c.PopulateL4PolicySetToCache(client[6], cloud)
 	}()
+	go func() {
+		if len(client) > 8 {
```
Comment (akshayhavile) [inline internal/cache/controller_obj_cache.go:119]: This `if loop` is not required. We already creating `avi client of length 9 at https://github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/blob/master/internal/cache/avi_ctrl_clients.go#L61

File: internal/nodes/avi_model_nodes.go [hunk 6]
```diff
@@ -1795,3 +1807,102 @@ type HostNamePathSecrets struct {
...
+			aviSG = append(aviSG, sg)
+		}
+	}
+	return aviSG
+}
+
+func (o *AviObjectGraph) GetAviStringGroupNodeByName(stringgroupName string) *AviStringGroupNode {
+	for _, model := range o.modelNodes {
+		sg, ok := model.(*AviStringGroupNode)
+		if ok && *sg.StringGroup.Name == stringgroupName {
```
Comment (akshayhavile) [inline internal/nodes/avi_model_nodes.go:1859]: Nit: `stringgroupName` --> `stringGroupName`

File: internal/nodes/avi_model_nodes.go [hunk 7]
```diff
@@ -27,6 +27,7 @@ import (
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/internal/lib"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
 
+	"github.com/vmware/alb-sdk/go/models"
```
Comment (DixitAakash) [inline internal/nodes/avi_model_nodes.go:30]: Looks like we have imported the same package twice and even using it with different names in the following code.

File: internal/cache/controller_obj_cache.go (hunk 5 above)
Comment (akshayhavile) [reply internal/cache/controller_obj_cache.go:119]: I should have elaborated the comment. Intention was to check the length of aviclient in code. If it is not as per our requirement, increase over there.

File: internal/rest/avi_obj_string_group.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019-2020 VMware, Inc.
```
Comment (DixitAakash) [inline internal/rest/avi_obj_string_group.go:2]: Please update the year.

File: internal/rest/dequeue_nodes.go [hunk 9]
```diff
@@ -1855,3 +1870,127 @@ func (rest *RestOperations) PkiProfileDelete(pkiProfileDelete []avicache.Namespa
...
+		utils.AviLog.Debugf("key: %s, msg: Executing rest for datascript %s", key, datascriptName)
+		utils.AviLog.Debugf("key: %s, msg: restops %v", key, rest_ops)
+		success, _ := rest.ExecuteRestAndPopulateCache(rest_ops, ds_key, avimodel, key, false)
+		if success {
+			utils.AviLog.Debugf("key: %s, msg: the Datascript added successfully: %s", key, cache_ds_node)
+		}
+	}
+}
+
+// func (rest *RestOperations) dataScriptDelete(ds_to_delete []avicache.NamespaceName, namespace string, rest_ops []*utils.RestOp, key string) []*utils.RestOp {
```
Comment (arihantg) [inline internal/rest/dequeue_nodes.go:1980]: should we remove commented code ?

File: internal/rest/avi_obj_string_group.go (hunk 8 above)
Comment (pkoshtavmware) [reply internal/rest/avi_obj_string_group.go]: Done

File: internal/nodes/avi_model_nodes.go (hunk 7 above)
Comment (pkoshtavmware) [reply internal/nodes/avi_model_nodes.go]: Changed

File: internal/nodes/avi_model_nodes.go (hunk 6 above)
Comment (pkoshtavmware) [reply internal/nodes/avi_model_nodes.go]: Changed

File: internal/cache/controller_obj_cache.go (hunk 5 above)
Comment (pkoshtavmware) [reply internal/cache/controller_obj_cache.go]: Reverted the check

File: internal/rest/dequeue_nodes.go (hunk 9 above)
Comment (pkoshtavmware) [reply internal/rest/dequeue_nodes.go]: Done

**Review by DixitAakash — DISMISSED** [review]
LGTM

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Review by arihantg — APPROVED** [review]

**Comment by pkoshtavmware** [comment]
FT run result:
Screenshot 2024-06-21 at 12 51 30 PM

**Review by DixitAakash — APPROVED** [review]

**Review by arihantg — DISMISSED** [review]

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Comment by pkoshtavmware** [comment]
build ako

**Review by arihantg — APPROVED** [review]

**Review by DixitAakash — APPROVED** [review]

**Pull request merged**